hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -0,0 +1,326 @@
+  object Missing extends Init[Nothing]
+  case class Start[S](s0: S) extends Init[S]
+
+  sealed trait Step[+A, +S]
+  object EOS extends Step[Nothing, Nothing]
+  case class Skip[S](s: S) extends Step[Nothing, S]
+  case class Yield[A, S](elt: A, s: S) extends Step[A, S]
+
+  def stepIf[A, S, X](k: Step[A, S] => Code[X], c: Code[Boolean], a: A, s: S): Code[X] =
+    c.mux(k(Yield(a, s)), k(EOS))
+
+  trait Parameterized[-P, +A] { self =>
+    type S
+    val stateP: ParameterPack[S]
+
+    // - 'step' must maintain the following invariant: step(..., emptyState, k) = k(EOS); it should
+    // hopefully be a cheap operation to compute this.
+    // - emptyState is unfortunately needed in some situations to get around the lack of "Option[T]"s
+    // being (easily) possible in bytecode.
+    def emptyState: S
+
+    def length(s0: S): Option[Code[Int]]
+
+    def init(
+      mb: MethodBuilder,
+      jb: JoinPointBuilder,
+      param: P
+    )(k: Init[S] => Code[Ctrl]): Code[Ctrl]
+
+    def step(
+      mb: MethodBuilder,
+      jb: JoinPointBuilder,
+      state: S
+    )(k: Step[A, S] => Code[Ctrl]): Code[Ctrl]
+
+    def map[B](f: A => B): Parameterized[P, B] =
+      contMap[B] { (a, k) => k(f(a)) }
+
+    def contMap[B](
+      f: (A, B => Code[Ctrl]) => Code[Ctrl],
+      setup: Code[Unit] = Code._empty,
+      cleanup: Code[Unit] = Code._empty
+    ): Parameterized[P, B] = new Parameterized[P, B] {
+      type S = self.S
+      val stateP = self.stateP
+      def emptyState = self.emptyState
+      def length(s0: S): Option[Code[Int]] = self.length(s0)
+      def init(mb: MethodBuilder, jb: JoinPointBuilder, param: P)(k: Init[S] => Code[Ctrl]): Code[Ctrl] =
+        self.init(mb, jb, param) {
+          case Missing => k(Missing)
+          case Start(s) => Code(setup, k(Start(s)))
+        }
+      def step(mb: MethodBuilder, jb: JoinPointBuilder, state: S)(k: Step[B, S] => Code[Ctrl]): Code[Ctrl] =
+        self.step(mb, jb, state) {
+          case EOS => Code(cleanup, k(EOS))
+          case Skip(s) => k(Skip(s))
+          case Yield(a, s) => f(a, b => k(Yield(b, s)))
+        }
+    }
+
+    def filterMap[B](f: (A, Option[B] => Code[Ctrl]) => Code[Ctrl]): Parameterized[P, B] = new Parameterized[P, B] {
+      type S = self.S
+      val stateP = self.stateP
+      def emptyState = self.emptyState
+      def length(s0: S): Option[Code[Int]] = None
+      def init(mb: MethodBuilder, jb: JoinPointBuilder, param: P)(k: Init[S] => Code[Ctrl]): Code[Ctrl] =
+        self.init(mb, jb, param)(k)
+      def step(mb: MethodBuilder, jb: JoinPointBuilder, state: S)(k: Step[B, S] => Code[Ctrl]): Code[Ctrl] =
+        self.step(mb, jb, state) {
+          case EOS => k(EOS)
+          case Skip(s) => k(Skip(s))
+          case Yield(a, s) => f(a, {
+            case None => k(Skip(s))
+            case Some(b) => k(Yield(b, s))
+          })
+        }
+    }
+  }
+
+  val missing: Parameterized[Any, Nothing] = new Parameterized[Any, Nothing] {
```

Comment:
Nitpicking, but I'd expect this to be called the `empty` stream, not `missing`.

Reply:
it's not really an empty stream -- its more like a "null"/"NA" stream.

Reply:
Is the distinction you're making that missing streams return `Missing` on `init`, while empty streams successfully `init`, but return `EOS` on the first `step`? Are those two cases ever treated differently?

Reply:
yes. if you convert a `Missing` stream to an array, you get `None`. if you convert an empty stream to an array, you get `[]`.

Reply:
Ah, that makes sense, thanks!